### DISCUSSION TITLE
Source generators for StyledProperty/DirectProperty/AttachedProperty boilerplate

### DISCUSSION THREAD

**Discussion by lucdem** [AvaloniaUI/Avalonia#14731] [Ideas]
I created a [proof of concept repo](https://github.com/lucdem/AvaSourceGenerators), but the idea is that it would be nice to have a code generator that can take this:

```C#
public partial class MyControl : TemplatedControl
{
    [AvaStyledProperty]
    private string myStr = string.Empty;
}
```
And generate this:

```C#
public partial class MyControl
{
        public static readonly StyledProperty<string> MyStrProperty = AvaloniaProperty.Register<MyControl, string>("MyStr");
        
        public string MyStr
        { 
	        get => GetValue(MyStrProperty);
	        set => SetValue(MyStrProperty, value);
        }
}
```

**Comment by jack775544** [comment]
I have used this library in the past as well: https://github.com/HavenDV/DependencyPropertyGenerator

**Comment by maxkatz6** [comment]
We are waiting until it will be possible to make partial properties. Source generators on fields make little sense with styled properties, as these are not in fact backed by fields.

**Reply by maxkatz6** [reply]
But it's good to have options to explore anyway. Something like this would also make more sense with DirectProperties, which do use fields.

**Reply by lucdem** [reply]
It adds a small memory overhead of an extra field in the control, but I think that's a price worth paying for the added convenience. It's true partial properties would be a much better solution, but who knows when that's going to actually happen.

**Reply by timunie** [reply]
well for a 3rd party library as you have it, yes 100 %. for build in solution, probably not as it may be very confusing.

**Comment by Basyras** [comment]
How does it look like now @maxkatz6?

Is this still planned or already in work or looking for contribution?

**Reply by maxkatz6** [reply]
@Basyras it's not currently in work, but any interest is welcomed.
We would need to go with specification first to review, comparing existing implementations (third party Avalonia and WinUI ones, for example).

**Comment by Athari** [comment]
https://github.com/jp2masa/Avalonia.PropertyGenerator by @jp2masa is pretty good.
